Sync profile and apps

G-Ray opened the pull request.

G-Ray committed:
Add sync-profile messages

G-Ray committed:
sync: Check the db is authorized

G-Ray committed:
Add public profile to localstorage

G-Ray committed:
Profile sync

G-Ray committed:
clear localstorage after tests

G-Ray committed:
sync-profile: Move reusable functions to utils

G-Ray committed:
remove comments

G-Ray committed:
Check that synced profile is equal

G-Ray committed:
Move sync-profile to lib/

G-Ray committed:
Define and use waitForSync function

G-Ray committed:
Add localKey when adding device

G-Ray committed:
Update device with app localKey when adding an app

G-Ray committed:
Update device test with localKey

G-Ray committed:
openProfile: Register current device

G-Ray committed:
Sync and authorize other devices apps

G-Ray committed:
Handle app sync logic inside in masq.js

G-Ray committed:
ReAdd requiredParametersDevice

G-Ray commented on src/lib/masq.js:
```diff
@@ -98,7 +98,7 @@ const dispatchMasqError = (errorCode) => {
 
 class Masq {
-  constructor () {
+  constructor (dbNameSuffix = '') {
```
Workaround in order to avoid conflicts between dbNames during tests, because each HyperDB instance must have its own storage

sujeebant replied:
It looks like dbNameSuffix is not used

G-Ray replied:
not anymore at least, thanks

G-Ray commented on test/sync-profile.test.js:
```diff
@@ -0,0 +1,182 @@
+const { dbExists } = common.utils
+const { genAESKey, exportKey } = common.crypto
+
+const waitForSync = async (db1, db2) => {
```
A function similar to this one could help us detect if a DB is synced with another one. We will move this to another file

G-Ray commented on src/lib/masq.js:
```diff
@@ -153,11 +155,28 @@ class Masq {
+
+    // register current device if it does not exist
+    const device = await this.getDevice()
+    if (!device) {
```
TODO: move this to a function. This should add the device once and for all, for each new device

G-Ray commented on test/sync-profile.test.js:
```diff
@@ -0,0 +1,182 @@
+
+    await new Promise((resolve) => { setTimeout(resolve, 5000) })
+
+    expect(masq.appsDBs['app2' + '-copy']._authorized).to.have.lengthOf(2)
```
this `copy` things need to be refactored

G-Ray replied:
I don't think it's that bad as it's only in the context of the tests. We absolutely need to replicate data between two different Indexeddb databases

sujeebant commented on src/lib/masq.js:
```diff
@@ -39,7 +39,7 @@ const STATES = {
 let STUN_TURN = []
 
-if (process.env.REACT_APP_REMOTE_WEBRTC === 'true') {
+if (process.env.REACT_APP_REMOTE_WEBRTC && process.env.REACT_APP_REMOTE_WEBRTC === 'true') {
```
Is it necessary ?

sujeebant commented on src/lib/masq.js:
```diff
@@ -153,11 +155,28 @@ class Masq {
+    if (!device) {
+      const { name, os } = DetectBrowser.detect()
+      await this.addDevice({
+        name: `${capitalize(name)} - ${os}`,
```
why not capitalize os ?

sujeebant commented on src/lib/masq.js:
```diff
@@ -606,11 +643,54 @@ class Masq {
+        for (let app of otherDevice.apps) {
+          if (!this.appsDBs[app.discoveryKey]) {
+            // Add app
+            await this._createDBAndSyncApp(app.id + '-copy', app.key)
```
Why create it with '-copy' ?

G-Ray replied:
Because in the context of the tests, replication will occur in the same browser and tab. Hence, we need to replicate between two different IndexedDB databases.

sujeebant commented on src/lib/sync-profile.js:
```diff
@@ -0,0 +1,111 @@
+    this.key = await importKey(Buffer.from(rawKey, 'base64'))
+
+    this.hub = signalhub(channel, this.hubUrl)
+    this.hub.on('error', () => {})
```
signalhub error should be handled

sujeebant commented on src/lib/sync-profile.js:
```diff
@@ -0,0 +1,111 @@
+    const { msg, id, key, publicProfile } = await decryptJSON(data, this.key)
+    debug('pullProfile received', msg, id, key)
+
+    if (!id || !key || !publicProfile || msg !== 'pushProfile') throw new Error('refused')
```
more specific errors should be used

sujeebant commented on src/lib/sync-profile.js:
```diff
@@ -0,0 +1,111 @@
+    const { msg: msg2 } = await decryptJSON(data, this.key)
+    debug('pullProfile received', msg2)
+
+    if (msg2 !== 'writeAccessGranted') throw new Error('refused')
```
more specific error should be used

leventdem replied:
yes, using MasqError, we already have a wrong message error : new MasqError(MasqError.WRONG_MESSAGE, `Unexpectedly received message with type ${msg2}`)

sujeebant commented on src/lib/sync-profile.js:
```diff
@@ -0,0 +1,111 @@
+    // this will block until the profile value is replicated
+    await this.db.getAsync('/profile')
+    // const test = await this.db.getAsync('/profile/protectedMK')
+    // console.log('test', test)
```
Is it sure that the data needed in /profile is also there when we can get /profile ?

sujeebant replied:
remove console.log

sujeebant commented on src/lib/sync-profile.js:
```diff
@@ -0,0 +1,111 @@
+    await this.masq.openProfile(id, 'pass')
+
+    // We have now the profile synced, stop replication.
+    // The user can now log in to start further replication of its profile and apps
```
// The user can now log in to start further replication of **the** profile and apps

sujeebant commented on src/lib/sync-profile.js:
```diff
@@ -0,0 +1,111 @@
+
+    data = await waitForDataFromPeer(this.peer)
+    const { msg, id, key, publicProfile } = await decryptJSON(data, this.key)
+    debug('pullProfile received', msg, id, key)
```
isn't it 'pullProfile' that is received ?

G-Ray replied:
I meant "pullProfile received" like "pullProfile (the function's name) received: [the message]"

sujeebant commented on src/lib/sync-profile.js:
```diff
@@ -0,0 +1,111 @@
+    const { msg: msg2, key } = await decryptJSON(data, this.key)
+    debug('pushProfile received', msg2, key)
+
+    if (!key || msg2 !== 'requestWriteAccess') throw new Error('msg not expected' + msg2)
```
more specific error should be used

G-Ray replied:
what would you prefer ?

sujeebant replied:
Specify the context e.g. function and message expected

sujeebant commented on src/lib/sync-profile.js:
```diff
@@ -0,0 +1,111 @@
+
+    data = await waitForDataFromPeer(this.peer)
+    const { msg: msg2, key } = await decryptJSON(data, this.key)
+    debug('pushProfile received', msg2, key)
```
'requestWriteAccess' is received ?

sujeebant commented on src/lib/sync-profile.js:
```diff
@@ -0,0 +1,111 @@
+    await this.db.getAsync('/profile')
+    // const test = await this.db.getAsync('/profile/protectedMK')
+    // console.log('test', test)
+    await this.masq.openProfile(id, 'pass')
```
why openProfile ?

sujeebant commented on test/masq.test.js:
```diff
@@ -289,7 +291,10 @@ describe('masq internal operations', function () {
-    expect(updatedDevices[0]).to.eql(device)
+    expect(devices[0].id).to.exist
+    expect(devices[0].localKey).to.exist
+    expect(devices[0].localKey).to.have.lengthOf(64)
```
move assertions regarding devices[0] before the updateDevice

sujeebant commented on test/masq.test.js:
```diff
@@ -371,6 +376,10 @@ describe('masq protocol', function () {
+  after(() => {
+    window.localStorage.clear()
+  })
+
```
Why clear the localStorage now ?

G-Ray replied:
to avoid any  side effects with other tests suites, because there are :beetle:

sujeebant commented on test/sync-profile.test.js:
```diff
@@ -0,0 +1,182 @@
+  return new Promise((resolve) => {
+    const checkVersion = () => {
+      db1.version((err, latestVersion) => {
+        if (err) { throw err }
```
this would throw inside the callback and never be thrown in checkVersion ?

G-Ray replied:
Yup, I should have called reject, fixed

sujeebant commented on test/sync-profile.test.js:
```diff
@@ -0,0 +1,182 @@
+
+    checkVersion()
+  })
+}
```
The purpose of the function doesn't seem clear.
return as soon as one of the dbs has a version equal to the one from the other db ?
Wouldn't it be sufficient to specifically check if db1 has a version equal to the version of db2 ?

G-Ray replied:
in any case, this would still be the same implementation right ?

sujeebant replied:
You are right, it seems like this is the right way to implement this test.

G-Ray committed:
Clean up code

G-Ray committed:
Devices: Add FloatingButton to open QRCodeModal

G-Ray committed:
Devices: add translation

G-Ray committed:
TextInput: Add padding when button is set

G-Ray committed:
QRCode set dynamic secure link

G-Ray committed:
QRCodeModal: Start pushProfile and sync animation

G-Ray committed:
sync: dispatch error on hub error

G-Ray committed:
Upgrade router

G-Ray committed:
Initial sync page to pull profile successfully

G-Ray committed:
Sync UI when pulling profile

G-Ray committed:
karma: Set NODE_ENV to test

G-Ray committed:
Fix duplicated apps DB and data when syncing

G-Ray committed:
throw and display error if already synced

G-Ray committed:
QRCodeModal: Get masq instance from actions

G-Ray committed:
masqs: Define standalone function syncApps

G-Ray committed:
syncApps: Use if/else for better readability

G-Ray committed:
Index dbs by dbName only

G-Ray committed:
prevent duplicated apps entries in devices

G-Ray committed:
Fetch and display synced devices

G-Ray committed:
Append count to device name

G-Ray committed:
AddProfile Modal

G-Ray committed:
Add SyncUrl to Login page

G-Ray committed:
Update package-lock

G-Ray committed:
lint

G-Ray committed:
lint

G-Ray committed:
Fix notification position

G-Ray committed:
SyncUrl modal

G-Ray committed:
Update fr.json

G-Ray committed:
Call Sync process from AddProfile

G-Ray committed:
Start SyncUrl from landing

G-Ray committed:
Update translations

G-Ray committed:
Update package-lock

G-Ray committed:
Fix syncing modal

G-Ray committed:
Display AddProfile modal when linking app

G-Ray committed:
Login: Set z-index to plus square

G-Ray committed:
SyncUrl: Update go back button text

G-Ray committed:
Devices: Set FloatingButton as fixed

G-Ray committed:
SyncUrl: Fix mobile style

G-Ray committed:
Sync: Fix fetchUsers call when sync is complete

G-Ray committed:
Add password modal during sync

G-Ray committed:
Login: remove unused Sync component

G-Ray committed:
Split pullProfile into two steps, handle canceling

G-Ray committed:
Workaround for chrome when cancelling sync

G-Ray committed:
Update i18n

G-Ray committed:
rm comments

G-Ray committed:
Add Scanner modal

G-Ray committed:
Scanner: Add crosshair

G-Ray committed:
Scanner: Update i18n

G-Ray committed:
Fix password logic

G-Ray committed:
Sync: Fix password check

G-Ray committed:
Use switch statement

G-Ray committed:
Scanner: Ask for media permission

G-Ray committed:
Remove scanner from app.jsx

G-Ray committed:
Add error message when sync link is invalid

leventdem committed:
Add QrCode onboarding (#162)

* first version

* update next and previous icons

* update typo with specific classes

* add a color

* add bullets (progress bar)

* add cursor pointer

* refactor

* small bug fix (props)

* fix active non active bullets color

* reduce width to compensate a margin in OnBoarding

* add synchronization method modal

* change behaviour of sync button (open SyncMethod modal)

* refactor css, now works with desktop and mobile

* small fixes (add cursor pointer...)

* add onBoarding link for SyncUrl modal

* Add logic for onboardings

Must distinguish two cases:
- on mobile: an additional modal is created with two different onboardings (link, qrcode)
- on desktop: there is no additional modal

On SyncUrl modal : add a link to copy onboarding

* remove lastStep (not used here)

* add color names instead of #Hex and remove text-align css property

* update color values with names from colors.scss

* add english version of text

* create two separate components for each onBoarding

* update Login to take into account the newly created onboarding components

* add animations for onboardings

* add animations into onBoardings components

* update main onboarding component with animations

* add max-width and refactor bullets

* update i18n

* fix fr.json, remove \ before ' character

* fix bug when there is no users (click on sync existing profile)

* rm comments and fix css

* remove log in button onClick action

* rm log

G-Ray committed:
Scanner: remove Space before title

G-Ray committed:
Display scanner when button is clicked

G-Ray committed:
AddProfile: align title

G-Ray committed:
Center title-modal typography

G-Ray committed:
Scan and replace location

G-Ray committed:
Fix modal display bug on mobile

G-Ray committed:
Move onboarding logic for copy to SyncUrl

G-Ray committed:
SyncDevice: Add pading on modal

G-Ray committed:
Button: Set min height

G-Ray committed:
Scanner: Handle error instead of asking for perms

G-Ray committed:
Decrease Signup padding

G-Ray committed:
Remove log

leventdem committed:
update click behaviour of already got a profile button for mobile (#164)

leventdem committed:
fix text, by default onSave transforms ' into \i which causes issue xith i18n

G-Ray committed:
Revert "Fix modal display bug on mobile"

This reverts commit f59080d76da83169ddfca2e030f70ceb8f73c16d.

G-Ray committed:
Modal: Remove padding-left and right on mobile

G-Ray committed:
Sync: Use publicProfile sent from peer

G-Ray committed:
Avatar: Set min-width and min-height

G-Ray committed:
Setting: move title

leventdem committed:
remove all vertical spaces and add space-around attribute

G-Ray committed:
Scanner: bind handleError

G-Ray committed:
Modal: Update padding on mobile

leventdem committed:
Small fixes (#165)

* update fr

* add bullets on last slide

* update fr

* add bullets on last slide

* update translation

* fix typo

G-Ray committed:
Sync: Wait until apps dbs are writable

G-Ray committed:
Handle end message to complete sync

G-Ray committed:
Add mobile header in modals, and simplify modal

G-Ray committed:
Increase Signup width

G-Ray committed:
Modal: remove extra margin on content

G-Ray committed:
Sync: Add onClose handler

leventdem committed:
Refresh profile (#168)

* add refresh user action

* add watcher and handle update of profile DB

* add uodatePublicProfile method and manage watcher

* change ws

* Change behaviour for profile watcher

- refactor to have a generic addWatcher method
- transform watchers object to array
- add generic destroyWatchers method

* Check if null or undefined for watcher

* fix not fedined watcher case

leventdem committed:
Check if during pull profile an existing profile has the same username  (#169)

* Add message error in QrCodeModal for existing username

* Add check existing profiles' usernames

* Add error message for exiting username

* rm log

* Add a hack for tests + replace filter by find

* add french translation

* add default error message + fr translation

* Add default message

* update isUsernameAlreadyTaken (add uppercase) + test

G-Ray committed:
Margin: Set default padding to 32

G-Ray committed:
Modal: Fix defaultProps

G-Ray committed:
Modal: Set padding 0 as important on mobile

G-Ray committed:
Modal: Set min-height on mobile header

G-Ray committed:
Sync: Fix incorrect link detection

G-Ray committed:
Update SyncUrl title

G-Ray committed:
SyncUrl : Remove extra go back button

G-Ray committed:
DeleteAppDialog: Remove app name from title

G-Ray committed:
mobileHeader only on some modals

G-Ray committed:
Signup: keyboard accessibility

G-Ray committed:
Sync: Close when on error

G-Ray committed:
Revert "SyncUrl : Remove extra go back button"

This reverts commit 03d864d35a2f78a2352b750c991ea3115e5c94c8.

G-Ray committed:
SyncUrl: Remove go back button on mobile

G-Ray committed:
Handle goBack from addProfile modal

G-Ray committed:
SyncUrl: Use form

sujeebant committed:
package.json, package-lock.json: Update signalhubws dependency

G-Ray committed:
Setting: wrap inputs within a form

G-Ray committed:
Modal: Set overflow to improve autofocus

G-Ray committed:
App check for profile watcher before destroy

G-Ray committed:
Settings: Add autoComplete properties

G-Ray committed:
PasswordEdit: Use form

G-Ray committed:
Signup: Remove handleKeyUp

G-Ray committed:
Wrap Login password field in a form

leventdem committed:
fix center of next button

leventdem committed:
Rebase timout (#180)

* add promise timeout method

* Close sw if user click on close modal or click outside modal

* Add getter for this.sw

* Update call to joinSecureChannel with timeout + custom error message

* Increase timeout delay

* Remove setstate, message will eventually be updated with the default case

* Remove getSW, not needed

* Update translations

* Transform QRCOdeModal into class component

* Small refactor

G-Ray committed:
Login: preventDefault of form action

leventdem committed:
update fr and en translation (#182)

sujeebant committed:
containers/Landing: change layout of landing page

G-Ray committed:
Handle sync url with redux

G-Ray committed:
Modals - fix margin and improve responsiveness (#181)

* Improve modals responsiveness

* Handle scanner error with an error msg

* Add translations

* Sync: remove extra spaces

* Modal: Different padding on mobile and desktop

* Card: remove width

* Update fr.json

G-Ray committed:
Sync: Silent fail if peer is offline when aborting

G-Ray committed:
Sync: listen on url change

leventdem committed:
Add migration for new release (#186)

* Add migration from version alpha to 0.16

    In the old version of masq alpha, the device entry in the profile
    db did not contain the apps info after an activation with maps.
    We therefore add :
    - _checkversion during openProfile
    - _addVersion for addProfile
    - _migration in case we need it

* update .env

* fix missing await for _addVersion

* update device entry with apps in a generic way

* rm old code before refactor

* Move migration to a specific file + according changes

* Fix import

* Rm check getDevice (already done in openProfile)

G-Ray committed:
Sync modals are not closable

G-Ray committed:
sync-profile: fix aborting process

leventdem committed:
fix font size and color for syncMethod modal (#188)

G-Ray committed:
Workaround to effectively delete db on app load

G-Ray committed:
Fix remoeProfile test

G-Ray committed:
TextField: Set min-height

G-Ray committed:
Remove log
